Add Git version control editor plugin

Xrayez opened the pull request:
![image](https://user-images.githubusercontent.com/17108460/124599243-1ed55b80-de6e-11eb-8aa7-b852979caf41.png)

This integrates Git as a version control backend directly using C++ modules rather than GDNative. Initial implementation based on https://github.com/godotengine/godot-git-plugin, and adds https://github.com/libgit2/libgit2 as a submodule to Goost repository.

Unfortunately, it's not possible to select the Git addon via editor interface manually, because the interface was implemented to work with scripts that implement VCS addons. Therefore, unlike GDNative version, the Git addon is initialized automatically here at runtime, without needing to manually to create and assign GDNative-like scripts, which is a nice byproduct.

Additionally, it's **not currently possible to use this plugin without patching the engine source**. If you'd like to test this PR, you'll need to compile Goost with `scons use_godot_patches=yes`, this will apply `misc/patches/editor_vcs_interface_virtual.patch`, but should be possible once godotengine/godot#50219 is backported to 3.x.

Some features are not currently available, like networking (some dependencies have to be satisfied for that), which would be needed for future `push/pull` functionalities, see godotengine/godot#39255.

Currently tested on Windows (MSVC, MinGW) and Linux.

Co-authored-by: @ChronicallySerious

# Why?

The main purpose of having this implemented as a built-in module rather than GDNative plugin is that we'd like to have out-of-the-box experience using VCS. Currently, it takes a lot of steps to start using GDNative-based plugin (downloading the library, copying GDNative scripts). Moreover, if you use Godot as a prototyping tool a lot, you may certainly benefit from having VCS throughout any project.

All you'll have to do is to download the editor and simply start using Git without having to setup GDNative-specific scripts, and it will work for all projects. Speaking of myself, I'd mainly find this kind of functionality useful for prototyping purposes, which allows me to keep the development history (oftentimes important at early development stages, in my experience, because I tend to trash code I think I won't need, but eventually some parts are still useful). Having to setup GDNative scripts for each new project goes against the idea of quick prototyping, for instance see my other repository https://github.com/godot-extended-libraries/godot-main.

Because Godot is highly picky about what kind of third-party software it uses license-wise and size-wise, it cannot allow to integrate https://github.com/libgit2/libgit2 because it's licensed under GPLv2 license. Fortunately for Goost, it's should be still safe to use for editor builds because it's released with linking exception in the license (even Microsoft uses libgit2 for its Git extensions for Visual Studio). This is also the reason why https://github.com/godotengine/godot-git-plugin still integrates libgit2, but there are other reasons why it got implemented as a plugin, of course.

## To-do (the v1.0 version)

- [x] Port godot-git plugin v1.x to C++ modules.
- [x] Add a button to initialize git repository manually at Version Control menu option.
- [x] Automatically enable git plugin for the project if git repository is already initialized at startup: godotengine/godot-git-plugin#40
- [x] Do not create initial commit: godotengine/godot-git-plugin#23.
- [x] Implement proper staging mechanism: godotengine/godot-git-plugin#54.

## May be fixable:
- [ ] godotengine/godot#50427.

Xrayez committed:
Add Git version control editor plugin

This integrates Git as a version control backend directly using C++ modules rather than GDNative.

Unfortunately, it's not possible to select the Git addon via editor interface manually, because the interface was implemented to work with scripts that implement VCS addons. Therefore, unlike GDNative version, the Git addon is initialized automatically here at runtime, without needing to manually to create and assign GDNative-like scripts.

The initial implementation is based on https://github.com/godotengine/godot-git-plugin.

Co-authored-by: Twarit Waikar <wtwarit@gmail.com>

Xrayez committed:
Document Git plugin

[ci skip]

Xrayez committed:
Fix `EditorVCSInterfaceGit` memory leak

twaritwaikar commented on modules/git/git_api.cpp:
```diff
@@ -0,0 +1,296 @@
+void EditorVCSInterfaceGit::_setup() {
+	// EditorVCSInterfaceGit singleton must NOT be assigned until `register_editor` is called.
+	// otherwise no docks will appear.
+	VersionControlEditorPlugin::get_singleton()->register_editor();
```
It seems what this function does is that it tries to replace the object instance which the editor uses to do all the VCS operations.

However, can this actually leak memory if you don't destroy the original `EditorVCSInterface` instance that was present there before assigning it to an instance of `EditorVCSInterfaceGit` from your module?

Xrayez replied:
I think this could happen. But initialization of `EditorVCSInterfaceGit` starts before any other plugin, and as far as I know, no VCS is initialized at editor startup even when using GDNative, referring to godotengine/godot#49309.

twaritwaikar replied:
I see, still, I don't think it would hurt to see if the singleton is existing, if it does then delete it and then assign it to an object of your own. The editor takes care of when to destroy the plugin anyway.

twaritwaikar reviewed (commented):
Also what happens when the user tries to add the official Git plugin with this module existing in the engine?

Btw, it's great that you got the v1.x of the plugin working with it and I hope you have seen the [v2.0](https://github.com/godotengine/godot-git-plugin/pull/47) that's been [in the works](https://github.com/godotengine/godot/pull/39255) which basically adds a lot of requested functionalities. It would be nice to have that, in case you decide to integrate that as well, of course!

Xrayez commented:
> Also what happens when the user tries to add the official Git plugin with this module existing in the engine?

I may try and see how this could affect GDNative plugins, but as far as I'm concerned, this feature may end up being experimental. Resolving memory leaks should likely be solvable eventually, if that's really going to be a problem.

As I said, VCS in Godot's editor could also be improved instead to handle instances coming from modules. I'm basically restricted to existing system that uses script instances to initialize and list VCS providers.

> Btw, it's great that you got the v1.x of the plugin working with it and I hope you have seen the [v2.0](https://github.com/godotengine/godot-git-plugin/pull/47) that's been [in the works](https://github.com/godotengine/godot/pull/39255) which basically adds a lot of requested functionalities. It would be nice to have that, in case you decide to integrate that as well, of course!

Yeah, this looks nice, although I'd like to resolve this first and see whether I'd need those features myself, it should be possible to make further enhancements afterwards. It would likely make sense to keep the implementation in sync, but some things may diverge, I think this is inevitable.

Xrayez committed:
Resolve `prce` link size errors in Git plugin

Xrayez committed:
Fix warnings in Git common and API sources

Xrayez committed:
Do not compile `git` module if `libgit2` is not checked out.

Xrayez committed:
Fix linux platform detection for Git plugin in SCsub

Xrayez committed:
Update compile instruction when using submodules

[ci skip]

Xrayez committed:
Prevent potential memory leak when registering new VCS backend

Xrayez committed:
Compile missing unix sources for Git plugin

Xrayez commented:
I've tested this on a native Linux machine now (Fedora), works as well. I'll assume that it will also work on macOS.

The only thing left is making it actually possible to override those endpoint proxy functions via C++, so I've submitted godotengine/godot#50219 patch which should be safe to merge and backport in Godot.

I have discovered other usability issues like forcing initialization/initial commit godotengine/godot-git-plugin#23, so those are also worth to resolve. It would probably make sense to make some things separate from the existing Godot's VCS UI on some level, because I obviously bypass a lot of things to make this work. Fortunately, I have access to both VCS plugin *and* editor to integrate this well enough.

Xrayez committed:
Override virtual entrypoints rather than bindings in `EditorVCSInterfaceGit`

Depends on godotengine/godot#50219 to be ported to Godot 3.x.

Xrayez committed:
Minor grammar and `nullptr` style fixes in Git plugin

[ci skip]

Xrayez committed:
Rename `libgit2` submodule

Xrayez committed:
Add `git` module to `disable_builtin.py`

Note that Git plugin is enabled by default, but this just maintains a list of modules which can be instantly disabled.

[ci skip]

Xrayez committed:
Update modules/README.md

[ci skip]

Xrayez committed:
Add a button to setup git repository manually

Also adds Git icon to Version Control menu option.
Introduces Git manager which reduces complexity of the Git interface itself.

The plugin is also automatically enabled if `.git` is found at project startup.

Xrayez commented:
I've decided to manage object instances separately from script instances:

![image](https://user-images.githubusercontent.com/17108460/125116771-e182ff00-e0f5-11eb-9bba-4ae878ab39b6.png)

The plugin is now automatically enabled if Git repository is detected at startup, which resolves godotengine/godot-git-plugin#40.

It would likely make sense to introduce an editor setting to prevent automatic initialization at startup, since GDNative users might still want to use GDNative implementation.

Xrayez committed:
Rename setup/shutdown in Git plugin to use verbs

Xrayez committed:
Add editor setting for initializing Git plugin at editor startup

Allows GDNative users to use GDNative implementation freely, without having to manually disable built-in implementation first.

Xrayez commented:
> It would likely make sense to introduce an editor setting to prevent automatic initialization at startup, since GDNative users might still want to use GDNative implementation.

![image](https://user-images.githubusercontent.com/17108460/125120429-1e052980-e0fb-11eb-92c9-d82adf864d35.png)

Xrayez committed:
Do not create empty initial commit automatically

Let users control the contents of initial commits.

Xrayez committed:
Update generation of .gitignore and .gitattributes for Git plugin

Use raw string literals, should be easier to update in the future.

Xrayez committed:
Set minimum size for "Version Control" diff bottom panel

See upstream PR: godotengine/godot#50341.

Xrayez committed:
Make it possible to configure username and email via editor settings

A novice might not have Git installed/configured, so we should provide a built-in way to setup minimal configuration via editor.

Xrayez committed:
Properly stage files using Git API instead of tracking files via editor

Xrayez commented:
I have resolved the issue with staging files as described in godotengine/godot-git-plugin#54. I'm still using the v1.0 API of Godot's VCS interface (v2.0 is still WIP in Godot).

We still have to wait for Godot to backport already merged godotengine/godot#50219.

That said, I'd say this PR is completed now.

**Note to self**: remove `misc\patches\editor_vcs_interface_virtual.patch` once godotengine/godot#50219 is backported to 3.x.

Xrayez committed:
Update libgit2 third-party info

[ci skip]

Xrayez committed:
Automatically refresh stage area upon Git setup

Xrayez committed:
Update Git plugin after window refocus

Shut down the plugin if Git repository could no longer be located, and refresh stage area upon window refocus automatically.

Xrayez committed:
Remove `editor_vcs_interface_virtual.patch`

Merged and backported in Godot now: godotengine/godot#50416.

[ci skip]

Xrayez committed:
Do not override minimum size for diff viewer

Because the same is done in Godot now, see merged and backported patch: godotengine/godot#50416.

Xrayez commented:
I've done a bunch of other usability stuff as seen above. I've also attempted to replicate VS Code functionality of Ctrl + Enter, but figured that it would be too hacky to do in Goost. Due to this, I created godotengine/godot#50432 which should hopefully be available in Godot 3.4.

godotengine/godot#50219 is already cherry-picked in godotengine/godot#50416 which is required for this feature to work in Goost.

Due to this, I see no reason why this cannot be merged now.

According to Godot contributors chat discussion, the push/pull features are unlikely to be backported to Godot 3.x, see https://chat.godotengine.org/channel/editor?msg=KaZBJ9MupK66tPhjC.

Xrayez commented:
This is the first PR which introduced a new *git* submodule.

For existing users, you'll need to `git pull` and `git submodule update --init --remote` in order to compile the git module. 😃
